### DISCUSSION TITLE
Self-Intersecting LineString Intersection with a Polygon

### DISCUSSION THREAD

**Discussion by PerziThomas** [NetTopologySuite/NetTopologySuite#545] [Q&A] [answered]
### Description
Hello everyone,
My goal is to process a LineString in such a way that I can get the coordinates in which the line crosses a polygons border to enter or leave it.
Right now I am handling it using the **.Intersection()** method (eg. linestring.Intersection(polygon)) and this has been working quite well for most of my test cases. Issues only arise when using a LineString which intersects itself, as Intersection seems to return a MultiLineString made out of each little piece of the original LineString.

> LINESTRING(-3.7666583061218257 40.3371359374876,-3.766623437404633 40.33712673718736,-3.7665751576423645 40.33713491523207,-3.7665805220603947 40.33720647308101,-3.7666609883308406 40.33720545082656,-3.7666516005992885 40.33716967191103,-3.766470551490784 40.337178872205385,-3.7665121257305145 40.337262697051926,-3.7666556239128113 40.33729234239951,-3.7666247785091405 40.33736594458554,-3.766477257013321 40.33737616710505,-3.766348510980606 40.33724225197699,-3.7666019797325134 40.33722794042086,-3.766744136810303 40.33724838550012,-3.766711950302124 40.33741501266499,-3.766941279172898 40.33747328096297,-3.7668702006340027 40.33755710544369,-3.7666422128677373 40.337527460212385,-3.7665483355522156 40.33749474821795,-3.7664490938186646 40.33745488045324)

> POLYGON((-3.7668675184249873 40.337045978942626,-3.7666797637939444 40.33766341926946,-3.764957785606384 40.337274964093865,-3.7651509046554557 40.33680472588907,-3.7668675184249873 40.337045978942626))

I added a polygon and a linestring to reproduce the behaviour, which produce the following result from the Intersection method:
> MULTILINESTRING ((-3.7666583061218257 40.3371359374876, -3.766623437404633 40.33712673718736, -3.7665751576423645 40.33713491523207, -3.7665780434352985 40.33717340983366), (-3.7665780434352985 40.33717340983366, -3.7665805220603947 40.33720647308101, -3.7666609883308406 40.33720545082656, -3.7666516005992885 40.33716967191103, -3.7665780434352985 40.33717340983366), (-3.7665780434352985 40.33717340983366, -3.766470551490784 40.337178872205385, -3.7664978049082563 40.33723382242341), (-3.7664978049082563 40.33723382242341, -3.7665121257305145 40.337262697051926, -3.7666556239128113 40.33729234239951, -3.7666247785091405 40.33736594458554, -3.766477257013321 40.33737616710505, -3.766348510980606 40.33724225197699, -3.7664978049082563 40.33723382242341), (-3.7664978049082563 40.33723382242341, -3.7666019797325134 40.33722794042086, -3.766744136810303 40.33724838550012, -3.766711950302124 40.33741501266499, -3.766746837860798 40.33742387695956), (-3.766710906394139 40.337536392426216, -3.7666422128677373 40.337527460212385, -3.7665483355522156 40.33749474821795, -3.7664490938186646 40.33745488045324))

I'd really appreciate any help with this, be it a way to somehow get the linestring back together or prevent the splitting from happening at all.

### Steps to Reproduce
Create a LineString and a Polygon and apply the .Intersection method, the result will be a MultiLineString with each LineString self-intersection creating more parts.

### System Configuration
Tell us about the environment where you are experiencing the bug:
- NetTopologySuite version: 2.3.0
- Other NetTopologySuite packages and versions: SqlServerBytes version 2.0.0
- Environment (Operating system, version and so on): Windows 10 Pro Version 21H1
- .NET Framework version: .NET 5

I thank everyone for reading all of this and offering potential help.

**Comment by FObermaier** [comment]
If you just want the points where the LineString enters or leaves the Polygon you should perform the Intersection between the LineString and the Polygon's boundaries (Exterior- and InteriorRings).

**Comment by PerziThomas** [comment]
First, thank you so much for the reply.

Second, while this is definitely an option, I'd additionally need to get the point before and after the intersection on the LineString, as I need to included data to calculate something afterwards.

Currently I am doing this by taking either the first/last point inside the polygon and then just picking the next/previous point, and calculating from there on.

Is there a fail safe way (the line going across the border and being closer than the last point inside) to get that last point inside the polygon on the LineString?

Best regards

**Comment by airbreather** [comment, marked as answer]
Feels like you need to loop over the `LineString` points in order, and then test every segment against the polygon's rings to be completely failsafe.  Checking each of the `LineString`'s defining points alone isn't guaranteed to work for all inputs, because a segment may enter and then exit the `Polygon` without any points in between.

I don't know the relevant APIs well enough to be able to write code for it right now, but **I think** it would be something like this:
```
if the first point in the LineString is outside the polygon:
    state = outsideThePolygon
else:
    state = insideThePolygon

For each segment in the LineString:
    keepChecking = true
    foundIntersections = { empty set } // clear every time you move onto a new segment of the LineString!
    while keepChecking:
        keepChecking = false
        If state is outsideThePolygon:
            If the segment intersects one of the polygon's exterior ring segments that's not in foundIntersections:
                Choose the intersecting segment that happens earliest along your LineString segment
                Do whatever business logic you have for a segment that enters a polygon
                state = insideThePolygon
                Add the intersecting segment to foundIntersections
                keepChecking = true
        Else if state is insideThePolygon:
            If the segment crosses any of the polygon's interior ring segments that's not in foundIntersections:
                Choose the intersecting segment that happens earliest along your LineString segment
                Do whatever business logic you have for a segment that exits a polygon
                state = outsideThePolygon
                Add the intersecting segment to foundIntersections
                keepChecking = true

**Comment by PerziThomas** [comment]
Thank you very much Joe! I will consider implementing such an approach!

Currently I have implemented a solution in which I split the LineString into points and then using `DistanceOp` to find the nearest point. Not completely fail safe, but might be good enough for what I am doing actually, I'll need to test that.

I thank you both very much for the assistance.

Best regards
